torch/nn/utils/parametrize.py

```diff
@@ -0,0 +1,375 @@
+from ..modules.module import Module
+from ..parameter import Parameter
+import functools
+import inspect
+
+
+class Parametrization(Module):
+    r"""A kind of Module that parametrizes a Parameter or a buffer in terms
+    of its forward function
+
+    After registering a Parametrization ``param`` on a tensor ``t`` within a module
+    ``module`` with :meth:`#register_parametrization`, ``module.t`` will be turn
+    into a property, and will return ``param(t)``, where ``t`` is the previous tensor
+
+    After registering a Parametrization, the parametrization will be accessible under
+    ``module.t_param``. The parametrized and unparametrized tensors may be accessed
+    through ``param.evaluate()`` and ``param.original()``
+
+    Before a parametrization is registered, it may be chained to others to compose them
+    via the ``chain`` method. A parametrization may be evaluated on a tensor ``t`` by
+    ``param.evaluate(t)``, even when it has not been registered
+
+    Parametrizations may be registered on parameters, buffers, and already parametrized
+    tensors
+
+        # Defining several parametrizations on the same tensor
+        module = nn.Linear(5,3)
+        p1 = MyParametrization()
+        p2 = MyParametrization()
+        # Register p1 on the Parameter "weight"
+        register_parametrization(module, "weight", p1)
+        # Register p2 on the parametrized Parameter "weight"
+        register_parametrization(module, "weight", p2)
+        assert(module.weight == p2(p1(module.weight_param.original())))
+
+    We can also first compose them and then register them. The following code is
+    equivalent to the previous one
+
+        module = nn.Linear(5,3)
+        # Now p2.evaluate computes p1 and then p2
+        p2.chain(p1)
```

Comment:
When you have multiple parameters, you can register them in a `ModuleList` and then implement the forward function as:
```
    def forward(self, Xs):
        return [p.evaluate(X) for p, X in zip(self.parametrizations, Xs)]
```
Funnily enough, I had already implemented a version of the lowrank example that you propose there, but with an SVD decomposition. In the case you propose, the parametrization should register an auxiliary parameter the size of the original vector, as the parametrization has more parameters than the original vector (it's a submersion).

The first and second points are very valid though. My reasoning to go with this design was the following: This chaining mechanism allows you, at any point, to evaluate the current value of the parametrization. I think of a parametrization as a RAII object over a buffer or a parameter, becoming the one that manages it. This way, if you have a chain of two parametrizations `p1` from R^n to R^k and `p2` from R^k to R^d over certain parameter `t` in R^n. In this setting, `p2` does not have to know that it is acting on a parametrized object, it can always just ask for the value of the parameter, and it will be given `p1(t)`, which lives in the correct space R^k, even though the original object doesn't.

This design also allows one to implement easily optimizations like the one currently being implemented in the pruning module, where if two pruning modules are put one after the other, their masks are merged into one by multiplying them. This could be done by simply overloading the `chain` method.

This is related to the point you made below of flattening this chain of parametrizations into something similar to a `Sequential` module. The problem with that is that each parametrization does not have a way to evaluate the current state of the tensor (which might be necessary to initialize the state of the parametrization, like the mask) and does not know anything about the previous parametrizations either (no optimization like that of pruning is possible).

Reply:
> When you have multiple parameters, you can register them in a ModuleList and then implement the forward function as:

Sorry, but I don't see how this applies to any of my points. Your example applies multiple independent parametrizations to multiple independent parameters, while I bring up issues about chaining parametrization on a single parameter.

> Funnily enough, I had already implemented a version of the lowrank example that you propose there, but with an SVD decomposition. In the case you propose, the parametrization should register an auxiliary parameter the size of the original vector, as the parametrization has more parameters than the original vector (it's a submersion).

What does that mean? Where does it have to register this "auxiliary parameter"? I think all of the state necessary to construct a parametrization should be held entirely in the `Parametrization` object. It should never leak its own details into other modules, but that means that you cannot assume that there's a single underlying tensor you'll be processing.

Reply:
> When you have multiple parameters, you can register them in a ModuleList and then implement the forward function as:

Sorry, but I don't see how this applies to any of my points. Your example applies multiple independent parametrizations to multiple independent parameters, while I bring up issues about chaining parametrization on a single parameter.

> Funnily enough, I had already implemented a version of the lowrank example that you propose there, but with an SVD decomposition. In the case you propose, the parametrization should register an auxiliary parameter the size of the original vector, as the parametrization has more parameters than the original vector (it's a submersion).

What does that mean? Where does it have to register this "auxiliary parameter"? I think all of the state necessary to construct a parametrization should be held entirely in the `Parametrization` object. It should never leak its own details into other modules, but that means that you cannot assume that there's a single underlying tensor you'll be processing.

> The first and second points are very valid though. My reasoning to go with this design was the following: ...

While I see what you're saying, I don't understand how does it apply to my first two points. Chaining parametrizations doesn't require them to be stateful (just think of them as composition of stateless functions!), and the `>>` syntax can be overloaded too. The only difference is that you'd be implementing `__rshift__` instead of `chain` in a subclass!

> This design also allows one to implement easily optimizations like the one currently being implemented in the pruning module, ...

I don't see how my proposed design would prevent this. You can inspect the parametrizations in `__rshift__` and can return different specialized composed forms if that can be done.

Reply:
> This is related to the point you made below of flattening this chain of parametrizations into something similar to a Sequential module. The problem with that is that each parametrization does not have a way to evaluate the current state of the tensor ... and does not know anything about the previous parametrizations either ....

Can you please elaborate on what do you mean when you say that "a parametrization doesn't have a way to evaluate the current state of the tensor"? Also, it's actaully good that it doesn't know anything about the next pamraetrizations, ebcause that enables composability! The optimization you've mentioned should be applied whenever `>>` is applied. If the two parametrizations can be composed into a single one, then return the optimized version. If they can't, wrap them in the `Sequential`-like container. If you get a `Sequential` and another parametrization to append, then consider if you can't fuse the last two together and return a new `Sequential` with an optimized last layer.

Reply:
On how to implement the low-rank layer: One of them would be to ask the low-rank parametrization to be applied on a tensor of size `(2, n)` and then implement
```
class LowRank(nn.Parametrization):
  def forward(self, t):
    a, b = t
    return a.unsqueeze(1) @ b.unsqueeze(0) # outer product
```
If the writer insists that the parametrization should be applied on a tensor, it is then when parameters in parametrizations come in to play. One would write:
```
class LowRankVector(nn.Parametrization):
    def __init__(self, n):
        super().__init__()
        self.b = nn.Parameter(torch.Tensor(1, n))

    def forward(self, a):
        return a.unsqueeze(1) @ self.b
```
Sometimes the latter option with parameters inside parametrizations is the only viable option if one wants to keep the same interface for different objects that implement different optimizations depending on the size of the input.

For this reason, I do not think that returning a new parametrization is a very good option, as doing so would mean that you have to copy that parameter. In the case of pruning, for example, it would mean that one has to copy the mask and having a duplicate mask as well. I do not think that we want to deep-copy a module every time we do `>>`. To resolve the possible ambiguity of the order in which `chain` works we could rename it to `compose`. That way `f.compose(g)` would read and be semantically as `f o g` in mathematical notation.

> Can you please elaborate on what do you mean when you say that "a parametrization doesn't have a way to evaluate the current state of the tensor

I completely agree that a parametrization should not know anything about the next parametrizations, but it should now about the previous ones. Now, consider the example of an adapter class, that takes a parametrization from R^n to R^n in the constructor and uses it to implement a parametrization form R^k to R^k via composing a function from R^k to R^n on the left and a function from R^n to R^k on the right. This case appears when trying to do orthogonal optimisation on rectangular matrices (R^{nxd}) based on optimization on squared orthogonal matrices (R^{nxn}). If one wants to do so, this adapter class has to prepend a parametrization to that going from R^k to R^n. When one has done so, the parametrization from R^n to R^n is ready to be used on elements of R^k. Furthermore, if it has functions that use the current value of the parametrization, it can just invoke `self.evaluate()` and, as it is aware of all the previous parametrizations that were chained to it, everything just works. If, on the other hand, we implement everything as a flat `Sequential`-like method, a parametrization that's in the middle could not evaluate all the parametrizations up to itself.

Reply:
> One of them would be to ask the low-rank parametrization to be applied on a tensor of size (2, n) and then implement

The assumption that you can pack all parameters of a parametrization into a single tensor is too strong. For starters, it's just not the right way to model this semantically. Secondly, what if you wanted to have a parametrization that would e.g. restrict a weight to be an affine transformation of another weight? You'd need to keep both the matrix and the bias, and doing that in a single tensor is super annoying.

> I do not think that we should support the possibility of applying parametrizations to two different parameters.

I don't see why we should disallow that, especially that both parameters and modules support that. It would seem very incosistent to me if you coudn't do that.

> Given that parametrizations can be stateful, I do not think that returning a new parametrization is a very good option, as doing so would mean that you have to copy that parameter.

I don't really follow your reasoning. The parametrization returned by `>>` can be a thin wrapper around existing parametrization objects, so there would be no need to reallocate them.

> To resolve the possible ambiguity of the order in which chain works we could rename it to compose.

I don't think this resolves the issue. Both "chain" and "compose" have the same meaning and don't really clarify the direction in which the parametrizations are composed. Modelling it after a mathematical notation doesn't really help as I personally often struggle to recall which function comes first in `f o g`.

> Furthermore, if it has functions that use the current value of the parametrization, it can just invoke self.evaluate() and, as it is aware of all the previous parametrizations that were chained to it, everything just works.

My question is why would it have functions that do that?

>  If, on the other hand, we implement everything as a flat Sequential-like method, a parametrization that's in the middle could not evaluate all the parametrizations up to itself.

I think that's a good thing, because it ensures compositionality! 

---

Ok, I think there are two ways to look at composition of parametrizations. One is that it's simply function composition, and they don't have to be aware of each other. This naturally leads to a `Sequential`-like construction to represent their composition. The other way is to consider them more like a stack. But then, I'd argue that the right way to do it is to apply the parametrizations _to parameters of existing parametrizations_. This way you get the nice property that you can still call `.evaluate()` and get the desired result (the "previous" parametrizations will be automatically invoked when their respective attributes are accessed).

Does that make sense? I still don't fully understand why you might want to evaluate a parametrization and have it be aware of the things that come before it so it would be helpful to know. Unfortunately I wasn't able to follow your example. Also, if we go down the second path, then I'm pretty sure that you shouldn't be able to pass in custom tensors to `.evaluate()`. A parametrization should then simply have `forward(self)` (i.e. no arguments) that produces the result whenever called (and possibly caches the result).

Reply:
Regarding your point of putting a parametrization that depends on several parameters, I have just implemented it so that one may write:
`P.register_parametrization(module, ["weight", "bias"], MyParametrization(), name="combined")`
where `MyParametrization()` is a parametrization whose forward takes two arguments. This way we allow to put parametrizations that take several inputs and return several outputs. Now the original tensors are not moved to the parametrization, but we just take a list of pointers to those tensors. The tensor will still be accessible under `_tensor_name`, that is, we prepend an underscore to its name.

I completely agree on the two ways of looking at parametrizations. Let me describe a more concrete, see if it makes sense. Assume that you have a function that goes from R^{n x n} to SO(n), where SO(n) is the space of the orthogonal matrices. This function reads in Pytorch:
```python
def f(X):
  X = X.tril(-1)
  X = X - X.t()      # X is now skew-symmetric
  return expm(X) # The matrix exponential of a skew-symmetric matrix is orthogonal
```
Now, assume that we want to parametrize a non-square matrix of shape `n x k` to have orthogonal columns. It turns out that, if we append a matrix of size `n x (n-k)` of zeros on the left and pass it to `f` we get an `n x n` matrix. If we then take the `k` first columns of this orthogonal matrix, we get a matrix whose columns are orthogonal. This construction makes quite a bit of mathematical sense, but the why is not important. The idea is that there is an embedding that has to be applied before the `f` and a projection that has to be applied afterwards. This idea may be abstracted into a class that roughly reads:
```python
class Fibration(P.Parametrization):
    def __init__(self, parametrization):
        self.parametrization = parametrization

    def embedding(self, X):
        raise NotImplementedError()

    def projection(self, X):
        raise NotImplementedError()

    def forward(self, X):
        X = self.embedding(X)
        X = self.parametrization.evaluate(X)
        X = self.projection(X)
        return X
```
Needless to say that represents a mathematical object of its own. Now, if we have a function inside `parametrization` that needs to make use of the evaluation of the current state of the parametrization chain, this implementation of `Fibration` will not work, as `self.parametrization` has not been registered anywhere.

On why does a parametrization need access to the parameters there are two examples. One is that of the initialization of parameters. In the RAII idea of a parametrization, they would take care of the initialization of their parameters, as they know how are they using them. For example, in the example of `f`, we are using `tril` to parametrize the skew-symmetric matrix, but we could also be using `triu`. As such, it is reasonable that the Parametrization that has `f` as its forward, is the one that exports methods to initialize the parameters that it manages. Here I have not explained why a parametrization would also need to evaluate the parametrization chain it is on, but I think that it follows reasonably from all this. I can continue this example to explain that if you still feel like it would help.

This RAII idea sort of enforces that a parametrization can only be registered in one parameter / tensor at a time. This idea is what the current implementation follows.

Reply:
Great, thanks for a more detailed write-up. To answer the first part (i.e. your example for why do we even need to be able to evaluate parametrizations on something else than its parameters), I think that we're mixing up two APIs that we really shouldn't. If you never need parameters, then you never should have made this thing a module in the first place. Ideally, you'd keep two APIs for parametrizations: one functional (where all parameters have to be passed in as arguments), and one `nn.Module` based, which takes care of state management for you, but generally defers to the stateless functional implementation in the `forward`. That's the pattern we already use for `nn` and we should continue it here.

> On why does a parametrization need access to the parameters 

I'm sorry, I can't follow your argument, because I'm not sure which part of my argumentation are you referring to. It would be very helpful if you could cite the relevant parts inline. Sorry 😕 

By the way do you think it might be helpful to have a higher bandwidth meeting some time (like a voice call)? I feel like we've been going back and forth a lot, and writing up those arguments and rereading all the previous messages takes a lot of time. Can you please send me an email describing your availability so that we can arrange for a time? I'm hoping that would allow us to find a good design much more quickly!